servers/rendering/renderer_rd/effects/metal_fx.mm

```diff
@@ -0,0 +1,214 @@
+		desc.outputTextureFormat = pf.getMTLPixelFormat(p_params.output_format);
+		desc.colorProcessingMode = MTLFXSpatialScalerColorProcessingModeLinear;
+		id<MTLFXSpatialScaler> scaler = [desc newSpatialScalerWithDevice:dev];
+		MFXSpatialContext *context = memnew(MFXSpatialContext);
+		context->scaler = scaler;
+		return context;
+	} else {
+		// Shouldn't be possible to get to this point.
+		CRASH_NOW_MSG("MetalFX spatial scaler are only available on macOS 13.0, iOS 16.0 or later.");
+	}
+}
+
+#pragma mark - Temporal Scaler
+
+MFXTemporalContext::~MFXTemporalContext() {}
+
+MFXTemporalEffect::MFXTemporalEffect() {}
+MFXTemporalEffect::~MFXTemporalEffect() {}
+
+MFXTemporalContext *MFXTemporalEffect::create_context(CreateParams p_params) const {
+	DEV_ASSERT(RD::get_singleton()->has_feature(RD::SUPPORTS_METALFX_TEMPORAL));
+
+	if (@available(macOS 13.0, iOS 16.0, *)) {
+		RenderingDeviceDriverMetal *rdd = (RenderingDeviceDriverMetal *)RD::get_singleton()->get_device_driver();
+		PixelFormats &pf = rdd->get_pixel_formats();
+		id<MTLDevice> dev = rdd->get_device();
+
+		MTLFXTemporalScalerDescriptor *desc = [MTLFXTemporalScalerDescriptor new];
+		desc.inputWidth = (NSUInteger)p_params.input_size.width;
+		desc.inputHeight = (NSUInteger)p_params.input_size.height;
+
+		desc.outputWidth = (NSUInteger)p_params.output_size.width;
+		desc.outputHeight = (NSUInteger)p_params.output_size.height;
+
+		desc.colorTextureFormat = pf.getMTLPixelFormat(p_params.input_format);
+		desc.depthTextureFormat = pf.getMTLPixelFormat(p_params.depth_format);
+		desc.motionTextureFormat = pf.getMTLPixelFormat(p_params.motion_format);
+		desc.autoExposureEnabled = NO;
+
+		desc.outputTextureFormat = pf.getMTLPixelFormat(p_params.output_format);
+
+		id<MTLFXTemporalScaler> scaler = [desc newTemporalScalerWithDevice:dev];
+		MFXTemporalContext *context = memnew(MFXTemporalContext);
+		context->scaler = scaler;
+
+		scaler.motionVectorScaleX = p_params.motion_vector_scale.x;
+		scaler.motionVectorScaleY = p_params.motion_vector_scale.y;
+		scaler.depthReversed = true; // Godot uses reverse Z per https://github.com/godotengine/godot/pull/88328
+
+		return context;
+	} else {
+		// Shouldn't be possible to get to this point.
+		CRASH_NOW_MSG("MetalFX temporal scaler is only available on macOS 13.0, iOS 16.0 or later.");
+	}
+}
+
+void MFXTemporalEffect::process(RendererRD::MFXTemporalContext *p_ctx, RendererRD::MFXTemporalEffect::Params p_params) {
+	CallbackArgs *userdata = &args[i];
+	userdata->src.id = RD::get_singleton()->get_driver_resource(RDC::DRIVER_RESOURCE_TEXTURE, p_params.src);
+	userdata->depth.id = RD::get_singleton()->get_driver_resource(RDC::DRIVER_RESOURCE_TEXTURE, p_params.depth);
+	userdata->motion.id = RD::get_singleton()->get_driver_resource(RDC::DRIVER_RESOURCE_TEXTURE, p_params.motion);
+	if (p_params.exposure.is_valid()) {
+		userdata->exposure.id = RD::get_singleton()->get_driver_resource(RDC::DRIVER_RESOURCE_TEXTURE, p_params.exposure);
```

Comment:
As you're reusing a previously used structure, you'll want to clear the contents of the exposure ID in the else branch here to indicate it's an empty ID.

Reply:
Good catch – fixed.